inconspiquous/transforms/xzs/commute.py

```diff
@@ -76,67 +74,90 @@ def match_and_rewrite(self, op1: qssa.DynGateOp, rewriter: PatternRewriter):
             return
 
-        if isinstance(op2.gate, HadamardGate):
-            new_op2 = qssa.GateOp(HadamardGate(), *op1.ins)
-            new_gate = XZOp(gate.z, gate.x)
-            new_op1 = qssa.DynGateOp(new_gate, *new_op2.outs)
-
-            rewriter.replace_op(op2, (new_op2, new_gate, new_op1))
+        if isinstance(op2.gate, CliffordGateAttr):
+            input_idx = use.index
+
+            x_prop = op2.gate.pauli_prop(input_idx, "X")
+            z_prop = op2.gate.pauli_prop(input_idx, "Z")
+
+            new_operands = list(op2.ins)
+            new_operands[input_idx] = op1.ins[0]
+
+            new_op2 = qssa.GateOp(op2.gate, *new_operands)
+            ops_to_insert: list[Operation] = []
+
+            false_const = arith.ConstantOp.from_int_and_width(0, 1)
+            false_const_needed = False
+            output_gates: list[qssa.DynGateOp | None] = []
+
+            for out_idx, ((x_from_x, z_from_x), (x_from_z, z_from_z)) in enumerate(
+                zip(x_prop, z_prop)
+            ):
+                apply_x: SSAValue
+                apply_z: SSAValue
+                needs_gate = False
+
+                if x_from_x and x_from_z:
+                    xor_x = arith.XOrIOp(gate.x, gate.z)
+                    apply_x = xor_x.result
+                    ops_to_insert.append(xor_x)
+                    needs_gate = True
+                elif x_from_x:
+                    apply_x = gate.x
+                    needs_gate = True
+                elif x_from_z:
+                    apply_x = gate.z
+                    needs_gate = True
+                else:
+                    apply_x = false_const.result
+                    false_const_needed = True
+
+                # Z component
+                if z_from_x and z_from_z:
+                    xor_z = arith.XOrIOp(gate.x, gate.z)
+                    apply_z = xor_z.result
+                    ops_to_insert.append(xor_z)
+                    needs_gate = True
+                elif z_from_x:
+                    apply_z = gate.x
+                    needs_gate = True
+                elif z_from_z:
+                    apply_z = gate.z
+                    needs_gate = True
+                else:
+                    apply_z = false_const.result
+                    false_const_needed = True
+
+                # Create XZ gate for this output
+                if needs_gate:
+                    xz_gate = XZOp(apply_x, apply_z)
+                    dyn_gate = qssa.DynGateOp(xz_gate, new_op2.outs[out_idx])
+                    ops_to_insert.extend([xz_gate, dyn_gate])
+                    output_gates.append(dyn_gate)
+                else:
+                    output_gates.append(None)
+
+            final_ops: list[Operation] = []
+            if false_const_needed:
+                final_ops.append(false_const)
+            final_ops.append(new_op2)
+            final_ops.extend(ops_to_insert)
+
+            new_outputs: list[SSAValue] = []
```

Comment:
It would be nice to just set the `output_gates` in the loop where you already have the correct index.

Reply:
@Gmin2 my comment wasn't the most helpful here, I meant that the variable `new_outputs` (not `output_gates`) could just be set in the loop, which would mean that there's never a need to have `output_gates` at all.